File: 08.DynamicProgramming/Leetcode/C++/446.ArithmeticSlicesII.cpp
#include <iostream>
#include <vector>
#include <unordered_map>

using namespace std;

int numberOfArithmeticSlices(vector<int>& A) {
    int alen = A.size();
    vector<unordered_map<int, int>> memo;

    for (int i = 1; i < alen; ++i) {
        for (int j = 0; j < i; ++j) {
            int difference = A[i] - A[j];
        }
    }

}

File: 08.DynamicProgramming/Leetcode/Python/446.ArithmeticSlicesII.py
class Solution:
    def numberOfArithmeticSlices(self, A):
        """
        :type A: List[int]
        :rtype: int
        """
        from collections import defaultdict
        alen = len(A)
        memo = [ defaultdict(int) for i in range(0, alen) ]
        result = 0
        for i in range(0, alen):
            for j in range(0, i):
                diff = A[i] - A[j]
                memo[i][diff] += 1
                if (diff in memo[j]):
                    memo[i][diff] += memo[j][diff]
                    result += memo[j][diff]


        #print(memo)
        return result;
